---
title: "R_Project_Analysis"
author: "Beatriz Pereira"
date: "2023-04-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# upload the necessary packages: 
```{r}
library(DESeq2) # differential gene expression analysis packet. 
library(dplyr) 
library(radiant)
library(tidyverse)
library(RColorBrewer)
library(pheatmap)
library(apeglm)
```

# load the data, row.names = 1 gets rid of 'x' header and makes miRNAs the row names: 
```{r}
countdata <- read.csv("GSE184723_miRNA_read-counts.csv.gz", header=T, row.names = 1)
```

# inspect the data: 
```{r}
type_countdata <- typeof(countdata) # we have a list
length_countdata <- length(countdata) #integer value
dim_countdata <- dim(countdata) # 1836 rows (our genes)
str_countdata <- str(countdata) # we have a data framem with 1836 obeservations and 16 variables. The gene names are characters, and the count data itself are integers. We have several biological replicates. 
```

# inspect RNA-seq count distribution: 
```{r}
ggplot(countdata) +
   geom_histogram(aes(x = AGAT_d.d10_4824), stat = "bin", bins = 200) + 
   xlim(-5, 500)  +
   xlab("Raw expression counts") +
   ylab("Number of genes")

# from these plots, we can see that there is a low number of counts associated with a large proportion genes, since there are so many genes with 0 expression. 

#Because this is count data, it should fit a negative binomial. Below, we are plotting the mean versus the variance of the data, for the AGAT deficient mice biological replicates [1-5]. 

mean_counts <- apply(countdata[, 1:5], 1, mean)
variance_counts <- apply(countdata[, 1:5], 1, var)
df <- data.frame(mean_counts, variance_counts)

ggplot(df) +
        geom_point(aes(x=mean_counts, y=variance_counts)) + 
        geom_line(aes(x=mean_counts, y=mean_counts, color="red")) +
        scale_y_log10() +
        scale_x_log10()
# this results in a negative binomial, so we can move onto our next steps. 
```

# making the coldata for DeSeq2
```{r}
# subset desired columns for analysis: 
countdata <- subset(countdata, select = c("AGAT_d.d10_4824", "AGAT_d.d_6_4619", "AGAT_d.d_7_4752",
                       "AGAT_d.d_8_4783", "AGAT_d.d_9_4822", "AGAT_WT_1_4617",
                       "AGAT_WT_2_4620", "AGAT_WT_3_4654", "AGAT_WT_4_4679",
                       "AGAT_WT_5_4692"))

# making the matrix - there are 15 total samples. 
coldata = matrix(c(1:10), ncol = 1, byrow = TRUE) 

# the condition is the genotype, so that is what we will want in the DeSeq object.
colnames(coldata) = "condition"

# make it a dataframe for easier management. 
coldata <- as.data.frame(coldata)

# assigning the rownames for coldata - they should all match the column names in countdata. 
rownames(coldata) <- c("AGAT_d.d10_4824", "AGAT_d.d_6_4619", "AGAT_d.d_7_4752",
                       "AGAT_d.d_8_4783", "AGAT_d.d_9_4822", "AGAT_WT_1_4617",
                       "AGAT_WT_2_4620", "AGAT_WT_3_4654", "AGAT_WT_4_4679",
                       "AGAT_WT_5_4692")

# adding in the genotypes - what the sample ID's correspond to.
coldata$condition = c('Deficient','Deficient','Deficient','Deficient','Deficient','Wildtype','Wildtype','Wildtype','Wildtype','Wildtype')

# check your dataframe. 
View(coldata)
```

# Making the DESeq object. 
```{r}
# make condition a factor. 
coldata$condition <- factor(coldata$condition) ## --- maybe delete --- 

# check that the order of column data (countdata) is consistent with the order of row data (coldata). DeSeq2 will not match these up for you and the analysis will be incorrect. 
rownames(coldata)
colnames(countdata) #samples order is consistent. 
all(rownames(coldata) %in% colnames(countdata)) # making sure that all samples in coldata are in countdata and vise versa. 
all(rownames(coldata) == colnames(countdata)) # rows and columns completely match. 

# making the DESeq object. 
dds <- DESeqDataSetFromMatrix(countData = countdata,
                              colData = coldata,
                              design = ~ condition)
# check on your DESeq object. 
dds
View(counts(dds))
```

# Pre-filtering low-expression miRNAs. Here, we are removing any count lower than 10, and keeping those in dds that are greater or equal to 10. 
```{r}
keep <- rowSums(counts(dds)) >= 10 
dds <- dds[keep,]
```

# Quality assessment: 
```{r}
# Transform counts for data visualization
rld <- rlog(dds, blind=TRUE) 
## blind = TRUE : transformation that is not biased to sample condition. rlog : returns a DESeqTransform object. 

# Plot PCA (principal component analysis)
plotPCA(rld, intgroup="condition") # this uses the top 500 most varibale genes. Biological samples should have similarities in clusters. 

# Hierarchial clustering: 
## Extract the rlog matrix from the object
rld_mat <- assay(rld)
## Compute pairwise correlation values
rld_cor <- cor(rld_mat)
## Plot heatmap
pheatmap(rld_cor) # our correlations range from 0.995 to 0.999. There don't seem to be extreme outliers. 
```

# Set control condition: 
```{r}
dds$condition <- factor(dds$condition, levels = c("Deficient","Wildtype"))
dds$condition <- relevel(dds$condition, ref = "Wildtype")
```

# DGE Analysis
```{r}
dds <- DESeq(dds) #this step automatically normalizes the data using median of ratios. 
# Dificient vs. Wildtype
res <- results(dds, contrast=c("condition","Deficient","Wildtype"))
# Applying False Discovery Rate. 
res$p_val_fdr <- p.adjust(res$pvalue, method = "fdr")
# make a tibble
res_tibble <- res %>%
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()
# get rid of FDR adjusted p-values below .05: 
pfdr_cutoff <- 0.05
res_tibble <- dplyr::filter(res_tibble, p_val_fdr < pfdr_cutoff) %>%
  dplyr::arrange(desc(log2FoldChange), 
          desc(p_val_fdr))
res_tibble
write.csv(as.data.frame(res_tibble), 
          file="deficient_control_results.csv")
# different results than supplementary table provided by research group. 
```

# Recreating the heatmap - not complete:
```{r}
library(ComplexHeatmap)

normalized_dds <- varianceStabilizingTransformation(dds, blind = FALSE)
rlog_out <- rlog(dds, blind = FALSE) # transformed with rlog and varianceStabilizingTransformation - rlog looked a little better. 

# quality check step (choosing which transformation): 
plotPCA(normalized_dds, intgroup = "condition")
plotPCA(rlog_out, intgroup = "condition") # use this one. 

# build a results table comparing miRNA expression between AGAT deficient and wildtype conditions: 
res_heatmap <- results(dds, contrast = c("condition","Deficient","Wildtype"))

# apply the FDR correction (multiple testing)
res_heatmap$p_val_fdr <- p.adjust(res_heatmap$pvalue, method = "fdr")

# get rid of na's in order to plot: 
sigs <- na.omit(res_heatmap)

# subset significant results, based on FDR: 
sigs <- sigs[sigs$p_val_fdr < 0.05,]

# make sigs into a dataframe: 
sigs <- as.data.frame(sigs)

mat <- counts(dds, normalized = TRUE)[rownames(sigs),]

# z scale: 
mat.z <- t(apply(mat, 1, scale))

# add desired column names: 
colnames(mat.z) <- c('AGAT-/-','AGAT-/-','AGAT-/-','AGAT-/-','AGAT-/-',
'wt','wt','wt','wt','wt')

# make out heatmap: 
mycol <- colorpanel(1000, low = "yellow", high = "red")

heatmap.2(mat.z, dendrogram = "column", key = TRUE,
          col = mycol, density.info="none", trace = "none",
          margins=c(5,10))
```

# Attempting to get their version (no FDR filtering): 
```{r}
head <- head(res[order(res$padj),], 20)
head <- as.data.frame(head)
mat.z <- t(apply(mat, 1, scale))
heatmap.2(mat.z, dendrogram = "column", key = TRUE,
          col = mycol, density.info="none", trace = "none",
          margins=c(5,10))
```

