---
title: "limmaAnalysis_Final"
author: "Jeniffer Perea-Lopez"
date: "2023-04-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---
title: "limmaAnalysis"
author: "Jeniffer Perea-Lopez"
date: "2023-04-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Upload the necessary packages
```{r}
library(limma)
library(edgeR)
```

#Reading in count-data
```{r}
miRNAcountdata <- read.csv("GSE184723_miRNA_read-counts.csv.gz", header=T, row.names = 1)
```

#Inspect the data
```{r}
type_countdata <- typeof(miRNAcountdata) # we have a list
length_countdata <- length(miRNAcountdata) #integer value
dim_countdata <- dim(miRNAcountdata) # 1836 rows (the miRNAs)
str_countdata <- str(miRNAcountdata) # we have a data frame with 1836 observations and 16 variables. The miRNA names are characters, and the count data itself are integers. We have several biological replicates. 
```

#Organizing sample information
```{r}
countdata <- data.frame(miRNAcountdata) #duplicate data for data processing
countdata <- countdata[,-c(11:15)] #remove columns with count-data for AGAT-deficient mice w/ hArg supplementation 
dge <- DGEList(counts=countdata) #create a DGEList object using edgeR package
dim(dge) #DGEList object contains a matrix of counts with 1836 rows associated with miRNAs and 10 columns associated with the individual samples
#adding sample name for further information on the type of mouse (i.e., genotype) and which mouse was analyzed
samplenames <- c('AGAT_deficient_mouse_10','AGAT_deficient_mouse_6',          'AGAT_deficient_mouse_7','AGAT_deficient_mouse_8',
'AGAT_deficient_mouse_9', 'AGAT_wildtype_mouse_1',
'AGAT_wildtype_mouse_2','AGAT_wildtype_mouse_3',
'AGAT_wildtype_mouse_4', 'AGAT_wildtype_mouse_5')
colnames(dge) <- samplenames
samplenames
#adding genotype (wild-type vs. AGAT knockout) of the mice
group <- as.factor(rep(c("AGATknockout", "Wildtype"), c(5,5)))
dge$samples$group <- group
#adding batch (sequencing lane) information
lane <- as.factor(rep(c("L008"), c(10)))
dge$samples$lane <- lane 

#Organizing miRNA annotations
miRNA <- row.names(dge) #extract miRNAs
miRNA <- data.frame(miRNA) #create dataframe with miRNAs
dge$miRNA <- miRNA #add miRNAs to DGEList-object
dge
```

#Normalization and filtering
```{r}
#Transformations from the raw scale
cpm <- cpm(dge) #convert raw counts to counts-per-million (CPM)
lcpm <- cpm(dge, log=TRUE) #convert raw counts to log2-CPM

L <- mean(dge$samples$lib.size) * 1e-6 #average library size in millions
M <- median(dge$samples$lib.size) * 1e-6 #median library size in millions
c(L, M)
summary(lcpm)

#Removing miRNAs that are lowly expressed 
table(rowSums(dge$counts==0)==10) #742 miRNAs have zero counts across all 10 samples
keep <- filterByExpr(dge, group=group) #remove rows that consistently have zero or very low counts
#By default, the filterByExpr function keeps genes with about 10 read counts or more in a minimum number of samples
#Since the paper seems to have filtered from 1836 miRNAs to 953 miRNAs, some parameters could be changed as to not filter out all genes that do not meet the standard (i,e., min.count = 10, total.min.count = 15, large.n = 10, min.prop = 0.7)
dge <- dge[keep,,keep.lib.sizes=FALSE]
dim(dge) #data was filtered from 1836 miRNAs to 226 miRNAs
lcpm.f <- cpm(dge, log=TRUE) #log2-CPM for filtered data 

#The figure shows the density of logCPM values for pre-filtered and post-filtered count-data
library(RColorBrewer)
nsamples <- ncol(dge)
col <- brewer.pal(nsamples, "Paired")
lcpm.cutoff <- log2(10/M + 2/L) #defines the logCPM threshold
par(mar = c(5, 5, 2, 2))
plotDensities(lcpm, col=col, main="", legend = "topright")
title(main="A.Density of logCPM for pre-filtered count-data")
abline(v=lcpm.cutoff, lty=3)
plotDensities(lcpm.f, col=col, main="", legend = "topright")
title(main="B. Density of logCPM for filtered count-data")
abline(v=lcpm.cutoff, lty=3)

#Normalizing gene distributions 
dge <- calcNormFactors(dge, method = "TMM") #apply scale normalization to RNA-seq read counts using the trimmed mean of M-values (TMM) normalizaton method 
dge$samples$norm.factors
```

#Differential expression: limma-trend 
```{r}
logCPM <- cpm(dge, log=TRUE, prior.count=3) #in the limma-trend approach, counts are converted to log-counts-per-million; prior.count is used here to damp down the variances of logarithms of low counts (i.e., ratio of the largest library size to the smallest is not more than about 3-fold)

#The figure shows the expression distribution of samples for un-normalized and normalized count-data
par(mar = c(2, 5, 2, 10))
boxplot(lcpm.f, col=col, main="", xaxt="n", ann=FALSE, xlim=c(0.5, 10.5), ylim = c(0,19))
title(main="A.Un-normalized count-data", ylab = "logCPM")
axis(1, las = 2, xpd = TRUE, at = NULL, labels = FALSE, tick = FALSE, line = NA)
mtext("Samples", side = 1, line = 1)
legend("topright", 
       inset = c(-0.48, 0),
       cex = 0.8,
       samplenames, 
       fill=col, 
       legend = samplenames, 
       bty="n", 
       title = "Samples", 
       xpd = TRUE)
boxplot(logCPM, col=col, main="", xaxt="n", ann=FALSE, xlim=c(0.5, 10.5), ylim = c(0,19))
title(main="B.Normalized count-data", ylab = "logCPM")
axis(1, las = 2, xpd = TRUE, at = NULL, labels = FALSE, tick = FALSE, line = NA)
mtext("Samples", side = 1, line = 1)
legend("topright", 
       inset = c(-0.48, 0),
       cex = 0.8,
       samplenames, 
       fill=col, 
       legend = samplenames, 
       bty="n", 
       title = "Samples", 
       xpd = TRUE)
```

#Differential expression analysis
```{r}
#Create a design matrix and contrasts
design <- model.matrix(~0+group) #set-up a design matrix with genotypes
colnames(design) <- gsub("group", "", colnames(design))
contr.matrix <- makeContrasts(KOvsWT = AGATknockout - Wildtype, levels = colnames(design)) #contrasts for pairwise comparison between genotypes

#Remove heteroscedascity from count data
par(mfrow=c(1,2))
v <- voom(dge, design, plot=TRUE) #transform count data to logCPM, estimate the mean-variance relationship to compute appropriate observational-level weights

#Fit linear models for comparison of interest
vfit <- lmFit(v, design) #fit linear model given a series of arrays
vfit <- contrasts.fit(vfit, contrasts=contr.matrix) #compute contrasts from linear model fit
efit <- eBayes(vfit, trend=TRUE) #compute moderated t-statistics, moderated F-statistic, and log-odds of differential expression by empirical Bayes moderation of the standard errors towards a common value, intensity-dependent trend is fitted to the prior variances
plotSA(efit, main="Final model: Mean-variance trend")

#Examine the number of differentially expressed (DE) genes with log-fold-changes significantly greater than 1
tfit <- treat(vfit, lfc=1) #use the treat method to calculate p-values from empirical Bayes moderated t-statistics with a minimum log-FC requirement

#Examine DE genes from smallest to largest adjusted p-value
topTreat <- topTreat(tfit, coef=1, n=Inf, adjust.method="BH") #table of top genes from a linear model fit, adjust p-Values using the False Discovery Rate (FDR) based Benjamini-Hochberg (BH) method
topTable <- topTable(efit, coef=1 , n=Inf, adjust.method="BH") #for results from eBayes, adjust p-Values using the BH method
```

#Using graphical representation of differential expression results 
```{r}
library(gplots)
KO.vs.WT <- topTreat #duplicate data
miRNAs <- row.names(KO.vs.WT)
KO.vs.WT$miRNAs <- miRNAs #add miRNAs to dataframe
KO.vs.WT.topgenes <- KO.vs.WT$miRNAs[1:33] #top 33 DE genes

logCPM2 <- data.frame(logCPM) #duplicate logCPM data and create a dataframe
logCPM2$miRNA <- row.names(logCPM2) #add miRNAs to dataframe
i <- which(logCPM2$miRNA %in% KO.vs.WT.topgenes) #returns the position which satisfies the given condition of the top 33 DE genes

#The figure shows a heatmap of logCPM values for top 33 genes DE in wild-type vs. AGAT-knockout
mycol <- colorpanel(1000, low = "yellow", high = "red") #low to high miRNA expresion is represented by a change of color from yellow to red 
heatmap.2(logCPM[i,], scale = "row", labRow = logCPM2$miRNA[i], labCol = group, col = mycol, trace = "none", density.info = "none", dendrogram = "column", key = TRUE, symkey = FALSE, keysize = 1.5)

eBayes_KO.vs.WT <- topTable #duplicate data
eBayes_miRNAs <- row.names(eBayes_KO.vs.WT)
eBayes_KO.vs.WT$eBayes_miRNAs <- eBayes_miRNAs #add miRNAs to dataframe
eBayes_KO.vs.WT.topgenes <- eBayes_KO.vs.WT$eBayes_miRNAs[1:33]
e <- which(logCPM2$miRNA %in% eBayes_KO.vs.WT.topgenes) #returns the position which satisfies the given condition of the top 33 DE genes from the eBayes method

#The figure shows a heatmap of logCPM values for top 33 genes DE in wild-type vs. AGAT-knockout using the eBayes method
mycol <- colorpanel(1000, low = "yellow", high = "red") #low to high miRNA expresion is represented by a change of color from yellow to red 
heatmap.2(logCPM[e,], scale = "row", labRow = logCPM2$miRNA[e], labCol = group, col = mycol, trace = "none", density.info = "none", dendrogram = "column", key = TRUE, symkey = FALSE, keysize = 1.5)
```